fs/osfs.py

```diff
@@ -361,90 +370,85 @@ def removedir(self, path):
     # Optional Methods
     # --------------------------------------------------------
 
+    # --- Type hint for opendir ------------------------------
+
     if False:  # typing.TYPE_CHECKING
 
         def opendir(self, path, factory=None):
             # type: (_O, Text, Optional[_OpendirFactory]) -> SubFS[_O]
             pass
 
-    def getsyspath(self, path):
-        # type: (Text) -> Text
-        sys_path = os.path.join(self._root_path, path.lstrip("/").replace("/", os.sep))
-        return sys_path
-
-    def geturl(self, path, purpose="download"):
-        # type: (Text, Text) -> Text
-        if purpose != "download":
-            raise NoURL(path, purpose)
-        return "file://" + self.getsyspath(path)
 
-    def gettype(self, path):
-        # type: (Text) -> ResourceType
-        self.check()
-        sys_path = self._to_sys_path(path)
-        with convert_os_errors("gettype", path):
-            stat = os.stat(sys_path)
-        resource_type = self._get_type_from_stat(stat)
-        return resource_type
+    # --- Backport of os.sendfile for Python < 3.8 -----------
+
+    def _check_copy(self, src_path, dst_path, overwrite=False):
+        # validate individual paths
+        _src_path = self.validatepath(src_path)
+        _dst_path = self.validatepath(dst_path)
+        # check src_path exists and is a file
+        if self.gettype(src_path) is not ResourceType.file:
+            raise errors.FileExpected(src_path)
+        # check dst_path does not exist if we are not overwriting
+        if not overwrite and self.exists(_dst_path):
+            raise errors.DestinationExists(dst_path)
+        # check parent dir of _dst_path exists and is a directory
+        if self.gettype(dirname(dst_path)) is not ResourceType.directory:
+            raise errors.DirectoryExpected(dirname(dst_path))
+        return _src_path, _dst_path
+
+
+    if sys.version_info[:2] < (3, 8) and sendfile is not None:
+
+        _sendfile_error_codes = frozenset({
+            errno.EIO,
+            errno.EINVAL,
+            errno.ENOSYS,
+            errno.ENOTSUP,
+            errno.EBADF,
+            errno.ENOTSOCK,
+            errno.EOPNOTSUPP,
+        })
+
+        @staticmethod
+        def _copy_shutil(src_path, dst_path):
+            return shutil.copy2(src_path, dst_path)
+
+        @classmethod
+        def _copy_sendfile(cls, src_path, dst_path):
+            sent = maxsize = 2**31 - 1
```

Comment:
Any particular reason to copy 2GB at a time?

Reply:
I think in the implementation I got that from they used that to ensure `sendfile` would be trying to send as much data as possible. It may be a better idea to replace that with an `os.fstat` call to get the filesize.